torch/_dynamo/utils.py

```diff
@@ -1173,14 +1173,14 @@ def visit(n: torch.fx.Node):
     if op == "call_module":
         nnmodule = tx.output.nn_modules[node.target]
 
-        if not is_lazy_module(nnmodule):
-            nnmodule = deepcopy_to_fake_tensor(nnmodule, tx.fake_mode)
-
-    if op == "call_module" and is_lazy_module(nnmodule):
-        assert nnmodule is not None
-        # In the case of a lazy module, we want to run
-        # the pre-hooks which initialize it
-        nnmodule(*args, **kwargs)
+        if is_lazy_module(nnmodule):
+            # In the case of a lazy module, we want to run
+            # the pre-hooks which initialize it
+            nnmodule._infer_parameters(nnmodule, args)
```

Comment:
what is the use-case of lazy module? is infer_parameters going to allocate big parameter tensors here (real tensors)?  Is that going to cause problems relative to how things behave in eager?

Reply:
The request comes from Meta Ads team, lazy module is initializing the module when the first iteration(by pre forward hook), and then the same as regular module since the second iteration. Yes, here we will allocate big parameters, but I think it's the same as eager mode.
Actually I was thinking to initialize the lazy module directly in fake mode, then no big parameter allocated. But I have concern if users are referring the module after compile, the module is fake tensor module, it caused discrepancy with eager mode.

Reply:
I guess i'm just not sure what the point of lazy module is if the weights can be initialized in memory.  I assumed there was some transform step that might reduce the size of the weight before it is finally initialized.

Also want to point out that torch.compile(lazy_model) appears to have a side-effect in your plan.  Generally, the model compiled by torch.compile is assumed not to be modified, but i guess lazy module is an exception.

Reply:
hmm, what if we just delete 'lazy module' support from dynamo, and instead, tell users they have to run their model once before compile? (also emit error from dynamo if compiling uninitialized lazy module)

Reply:
> Also want to point out that torch.compile(lazy_model) appears to have a side-effect in your plan. Generally, the model compiled by torch.compile is assumed not to be modified, but i guess lazy module is an exception.

 This is correct, but it seems this is the character of lazy module, because even in eager mode, the lazy module would be materialized after the first iteration, which is sort of side-effect.

 > hmm, what if we just delete 'lazy module' support from dynamo, and instead, tell users they have to run their model once before compile? (also emit error from dynamo if compiling uninitialized lazy module)

I'm ok with this, but I think my solution is to let dynamo to run the lazy module once during the compilation. I think my proposal is better for [the case like the unit test I added](https://github.com/pytorch/pytorch/pull/97946/files#diff-987e2ef5d2b9264b29d811f59a3fed2c98657ed05cd20f6b81ddb1d5331db406R470). In this example, the module users compiling is not lazy module, but it has submodule who are lazy. If the module architecture is complicated, users may get confused where the error is from. But I don't have strong opinion on this.

Reply:
Reading the [doc for LazyModuleMixin](https://pytorch.org/docs/stable/generated/torch.nn.modules.lazy.LazyModuleMixin.html) I realize the point of it is to initialize the parameters based on shapes from the first input.  Note this section 

"After construction, networks with lazy modules should first be converted to the desired dtype and placed on the expected device. This is because lazy modules only perform shape inference so the usual dtype and device placement behavior applies. The lazy modules should then perform “dry runs” to initialize all the components in the module. These “dry runs” send inputs of the correct size, dtype, and device through the network and to each one of its lazy modules. After this the network can be used as usual."

I think your PR is mostly in line with how users would expect this to work, except does it correctly handle changing the model's dtype/device _after_ the first iteration?

Also, I think LazyModules only work lazily one time, so if you feed fresh inputs of an incompatible shape it should error instead of working.  I think your dynamo PR does this part correctly, since dynamo guards would fail and attempt to recompile, but the recompile should also fail due to incompatible shapes. But you should add a test for this case and make sure the error is cleanly reported instead of hard to debug.

Reply:
> I think your PR is mostly in line with how users would expect this to work, except does it correctly handle changing the model's dtype/device after the first iteration?

This is a great point! If the model's dtype/device changes, it triggers recompilation. During recompilation, this module is not lazy anymore, I need to correctly handle this case and add a unit test.

> Also, I think LazyModules only work lazily one time, so if you feed fresh inputs of an incompatible shape it should error instead of working. I think your dynamo PR does this part correctly, since dynamo guards would fail and attempt to recompile, but the recompile should also fail due to incompatible shapes. But you should add a test for this case and make sure the error is cleanly reported instead of hard to debug.

Yes, dynamo's behavior are the same as eager. I'll add a unit test too.